pkg/operator/operator.go

```diff
@@ -241,76 +273,237 @@ func (k *Kontroller) withLeaderElection() error {
+// before-reboot=true label. This is considered the beginning of the reboot
+// process from the perspective of the update-operator. It will only mark
+// nodes with this label up to the maximum number of concurrently rebootable
+// nodes as configured with the maxRebootingNodes constant.
+// If there is an error getting the list of nodes or updating any of them, an
+// error is immediately returned.
+func (k *Kontroller) markBeforeReboot() error {
+	nodelist, err := k.nc.List(v1meta.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("Failed listing nodes: %v", err)
 	}
 
-	// Verify no nodes are still in the process of rebooting to avoid rebooting N > maxRebootingNodes
+	// find nodes which are still rebooting
 	rebootingNodes := k8sutil.FilterNodesByAnnotation(nodelist.Items, stillRebootingSelector)
-	glog.V(6).Infof("Found %+v rebooting nodes", rebootingNodes)
+	// nodes running before and after reboot checks are still considered to be "rebooting" to us
+	beforeRebootNodes := k8sutil.FilterNodesByRequirement(nodelist.Items, beforeRebootReq)
+	rebootingNodes = append(rebootingNodes, beforeRebootNodes...)
+	afterRebootNodes := k8sutil.FilterNodesByRequirement(nodelist.Items, afterRebootReq)
+	rebootingNodes = append(rebootingNodes, afterRebootNodes...)
+
+	// Verify the number of currently rebooting nodes is less than the the maximum number
 	if len(rebootingNodes) >= maxRebootingNodes {
-		glog.Infof("Found %d (of max %d) rebooting nodes; waiting for completion", len(rebootingNodes), maxRebootingNodes)
 		for _, n := range rebootingNodes {
 			glog.Infof("Found node %q still rebooting, waiting", n.Name)
 		}
-		return
+		glog.Infof("Found %d (of max %d) rebooting nodes; waiting for completion", len(rebootingNodes), maxRebootingNodes)
+		return nil
 	}
 
+	// find nodes which want to reboot
 	rebootableNodes := k8sutil.FilterNodesByAnnotation(nodelist.Items, wantsRebootSelector)
+	rebootableNodes = k8sutil.FilterNodesByRequirement(rebootableNodes, notBeforeRebootReq)
+
+	// Don't even bother if rebootableNodes is empty. We wouldn't do anything anyway.
 	if len(rebootableNodes) == 0 {
-		return
+		return nil
 	}
 
+	// find the number of nodes we can tell to reboot
 	remainingRebootableCount := maxRebootingNodes - len(rebootingNodes)
-	// pick N of the eligible candidates to reboot
+
+	// choose some number of nodes
 	chosenNodes := make([]*v1api.Node, 0, remainingRebootableCount)
 	for i := 0; i < remainingRebootableCount && i < len(rebootableNodes); i++ {
 		chosenNodes = append(chosenNodes, &rebootableNodes[i])
 	}
 
+	// set before=reboot=true for the chosen nodes
 	glog.Infof("Found %d nodes that need a reboot", len(chosenNodes))
-	for _, node := range chosenNodes {
-		k.markNodeRebootable(node)
+	for _, n := range chosenNodes {
+		err = k.mark(n.Name, constants.LabelBeforeReboot, k.beforeRebootAnnotations)
+		if err != nil {
+			return fmt.Errorf("Failed to mark node for before reboot checks: %v", err)
```

Comment:
Can we just be consistent in the logs and say "annotating node" or "labeling node". I'm mainly objecting to "mark" since it doesn't mean anything to a Kubernetes user